**Tuketal2017**\
*Landscape of X chromosome inactivation across human tissues*\
[https://doi.org/10.1038/nature24265](https://www.nature.com/articles/nature24265#citeas)\
-Suppl.Table.1.csv: combined cotton et al (2015) and carrel/willard (2005) escape calls
--note: to convert suppl.table.1.csv to the proper genome build (hg38), the chromosome positions\
were compiled, and uploaded to [NCBI Genome Remapping Service](https://www.ncbi.nlm.nih.gov/genome/tools/remap#)\
The source genome was hg19, the converted genome was hg38.\
The positions were then remapped to fit the GEUVADIS escape calls.


**Nelsonetal2015**\
*The support of human genetic evidence for approved drug indications*\
[https://doi.org/10.1038/ng.3314](https://www.nature.com/articles/ng.3314)\
-nelson_supp_dataset1: "Supplementary Data Set 1: GWASdb entries with MeSH terms mapped for each trait and genes annotated as described in the Online Methods."\
-nelson_supp_dataset2: "Supplementary Data Set 2: Reduction of the genetic association data to a single row per gene and trait as used for most analyses described."


**Meritxelletal2020**\
*The impact of sex on gene expression across human tissues*\
[DOI: 10.1126/science.aba3066](https://science.sciencemag.org/content/369/6509/eaba3066)\
-aba3066-Table-S3: Properties of sex-biased genes: For X-linked sex-biased genes, previously reported status of escape from X-inactivation (XCI). Sex prediction properties of X-linked and autosomal genes.


**KatsirLinial2019**\
*Human genes escaping X-inactivation revealed by single cell expression data*\
[DOI: 10.1186/s12864-019-5507-6](https://doi.org/10.1186/s12864-019-5507-6)
A single RNA-seq study on two cell types: fibroblast (n=104) & lymphoblast (n=104)
-table_s3.xlsx: "A list of informative SNPs (iSNPs) along with their labeling on ChrX, Chr17, all autosomal chromosomes and imprinted genes from 101 fibroblast single cells. (XLSX 60374 kb)"
-table_s3_mod.xlsx: Same as table_s3.xlsx but with the description row (first row) removed from sheet 'ChrX'.
--note: to convert table_s3 to the proper genome build (hg38), the chromosome positions\
were compiled, and uploaded to [NCBI Genome Remapping Service](https://www.ncbi.nlm.nih.gov/genome/tools/remap#)\
The source genome was hg19, the converted genome was hg38.\
-table_s5.xlsx: "A list of informative SNPs (iSNPs) along with their labeling on ChrX, Chr17 and autosomes from single cell lymphoblasts and Pool100 lymphoblasts. (XLSX 5482 kb)"